src/js/structs/List.js

```diff
@@ -74,15 +75,35 @@ module.exports = class List {
    */
   combine(list) {
     let myItems = this.getItems();
-    let newItems = list.reduceItems(function (items, item) {
-      if (myItems.indexOf(item) === -1) {
-        items.push(item);
-      }
+    let newItems = list.getItems();
+
+    // We pre-allocate the result array with the worst possible scenario
+    let combinedItems = new Array(myItems.length + newItems.length);
 
-      return items;
-    }, []);
+    // Instead of using the O(n²) approach of iterating over `newItems` and
+    // checking if each item exist on `myItems`, we use a trick that brings the
+    // complexity down to O(n) (well, O(2n) to be precise).
+
+    // We first generate a unique object for the current combine operation
+    // and we mark all the objects of `myItems` with it, declaring them as used.
+    let comparisonTag = {};
+    for (let i=0, l=myItems.length; i<l; ++i) {
```

Comment:
Let's use more descriptive names for `i`, `j` and `l` as you use them not only within the loop and it probably helps people to better understand this method.